opencsp/app/sofast/lib/ProcessSofastFixed.py

```diff
@@ -183,94 +161,201 @@ def generate_geometry(self, blob_index: BlobIndex, mask_raw: np.ndarray) -> dict
+        # Process optic geometry (find mask corners, etc.)
+        (
+            self.data_geometry_general,
+            self.data_image_processing_general,
+            self.data_geometry_facet,  # list
+            self.data_image_processing_facet,  # list
+            self.data_error,
+        ) = pr.process_multifacet_geometry(
+            self.data_facet_def,
+            self.data_ensemble_def,
+            mask_raw,
+            self.measurement.v_measure_point_facet,
+            self.orientation,
+            self.camera,
+            self.measurement.dist_optic_screen,
+            self.params.geometry,
+            self.params.debug_geometry,
+        )
 
-    def process_single_facet_optic(self, surface: Surface2DAbstract) -> None:
-        """Processes single facet optic. Saves data to self.data_slope_solver
+        kwargs_list = []
+        for idx_facet in range(self.num_facets):
+            # Get pixel region of current facet
+            loop = self.data_image_processing_facet[idx_facet].loop_facet_image_refine
+
+            # Get image points and blob indices
+            pts_image, pts_index_xy = blob_index.get_data_in_region(loop)
+
+            # Define optic orientation w.r.t. camera
+            rot_facet_ensemble = self.data_ensemble_def.r_facet_ensemble[idx_facet]
+            rot_ensemble_cam = self.data_geometry_general.r_optic_cam_refine_2
+            rot_facet_cam = rot_ensemble_cam * rot_facet_ensemble
+
+            v_cam_ensemble_cam = self.data_geometry_general.v_cam_optic_cam_refine_3
+            v_ensemble_facet_ensemble = self.data_ensemble_def.v_facet_locations[idx_facet]
+            v_ensemble_facet_cam = v_ensemble_facet_ensemble.rotate(rot_ensemble_cam)
+            v_cam_facet_cam = v_cam_ensemble_cam + v_ensemble_facet_cam
+
+            u_cam_measure_point_facet = self.data_geometry_facet[idx_facet].u_cam_measure_point_facet
+
+            # Get screen/camera poses
+            rot_cam_facet = rot_facet_cam.inv()
+            rot_facet_screen = self.orientation.r_cam_screen * rot_facet_cam
+            rot_screen_facet = rot_facet_screen.inv()
+
+            v_facet_cam_facet = -v_cam_facet_cam.rotate(rot_cam_facet)
+            v_cam_screen_facet = self.orientation.v_cam_screen_cam.rotate(rot_cam_facet)
+            v_facet_screen_facet = v_facet_cam_facet + v_cam_screen_facet
+
+            # Calculate xyz screen points
+            v_screen_points_screen = self.fixed_pattern_dot_locs.xy_indices_to_screen_coordinates(pts_index_xy)
+            v_screen_points_facet = v_facet_screen_facet + v_screen_points_screen.rotate(rot_screen_facet)
+
+            # Calculate active pixel pointing
+            u_pixel_pointing_cam = self.camera.vector_from_pixel(pts_image)
+            u_pixel_pointing_facet = u_pixel_pointing_cam.rotate(rot_cam_facet).as_Vxyz()
+
+            # Update debug data
+            self.params.debug_slope_solver.optic_data = self.data_facet_def[idx_facet]
```

Comment:
Should this be happening all the time or only when debug_active is true? If this is a reference copy rather than a resource-intensive copy, please disregard.

Reply:
It doesn't necessarily have to be done every time, but it is just a resource copy. If you'd rather have it only during active debugging, let me know and I can update.